### ISSUE TITLE
Missing `close_notify` is silently ignored

### ISSUE THREAD

**Issue by vilgotf** [sfackler/rust-openssl#1803]
Improper TLS terminations result in an unexpected EOF error, which, prior to OpenSSL 3.0 would set an errno of 0 (see BUGS section in https://www.openssl.org/docs/man1.1.1/man3/SSL_get_error.html). Since then OpenSSL now returns a `SSL_ERROR_SSL`, however, this is not caught by rust-openssl which erroneously returns `Ok(0)`. Relevant piece of code: https://github.com/sfackler/rust-openssl/blob/78c241f2716320f509f284261ce04812c3751b4f/openssl/src/ssl/mod.rs#L3691-L3709
Note that rust-openssl also does not register the error inside `InnerError` (`Error`'s cause field is `None`).

Somewhat related to tokio-rs/tls#128

**Comment by sfackler** [comment]
You can use `ssl_read` if you want more direct control over error handling. I would expect that changing the behavior here to return an error no a missing close_notify would cause fairly enormous amounts of breakage downstream.

**Comment by vilgotf** [comment]
I don't use rust-openssl directly, but rather through (tokio-)tls-native so I'm somewhat stuck with things as is (though I see you're the owner of tls-native too).

**Comment by ctz** [comment]
> I would expect that changing the behavior here to return an error no a missing close_notify would cause fairly enormous amounts of breakage downstream.

Yeah, that's not an ideal situation. However, as it currently stands I think conditionally swallowing these errors can only happen at higher levels from this crate. As an example: https://github.com/hyperium/h2/pull/743/files -- this is logic that will never be called when used on top of rust-openssl.

Unconditionally swallowing this error opens downstream users to truncation attacks; the impact of which is difficult to generalise.
